src/Lean/Parser/Basic.lean

```diff
@@ -422,379 +433,63 @@ def withResultOf (p : Parser) (f : Syntax → Syntax) : Parser := {
 
 def mkEmptySubstringAt (s : String) (p : String.Pos) : Substring := {
   str := s, startPos := p, stopPos := p
 }
 
-private def rawAux (startPos : String.Pos) (trailingWs : Bool) : ParserFn := fun c s =>
-  let input   := c.input
-  let stopPos := s.pos
-  let leading := mkEmptySubstringAt input startPos
-  let val     := input.extract startPos stopPos
-  if trailingWs then
-    let s        := whitespace c s
-    let stopPos' := s.pos
-    let trailing := { str := input, startPos := stopPos, stopPos := stopPos' : Substring }
-    let atom     := mkAtom (SourceInfo.original leading startPos trailing (startPos + val)) val
-    s.pushSyntax atom
-  else
-    let trailing := mkEmptySubstringAt input stopPos
-    let atom     := mkAtom (SourceInfo.original leading startPos trailing (startPos + val)) val
-    s.pushSyntax atom
-
-/-- Match an arbitrary Parser and return the consumed String in a `Syntax.atom`. -/
-def rawFn (p : ParserFn) (trailingWs := false) : ParserFn := fun c s =>
-  let startPos := s.pos
-  let s := p c s
-  if s.hasError then s else rawAux startPos trailingWs c s
-
-def chFn (c : Char) (trailingWs := false) : ParserFn :=
-  rawFn (satisfyFn (fun d => c == d) ("'" ++ toString c ++ "'")) trailingWs
-
-def rawCh (c : Char) (trailingWs := false) : Parser := {
-  fn := chFn c trailingWs
-}
-
-def hexDigitFn : ParserFn := fun c s =>
-  let input := c.input
-  let i     := s.pos
-  if h : input.atEnd i then s.mkEOIError
-  else
-    let curr := input.get' i h
-    let i    := input.next' i h
-    if curr.isDigit || ('a' <= curr && curr <= 'f') || ('A' <= curr && curr <= 'F') then s.setPos i
-    else s.mkUnexpectedError "invalid hexadecimal numeral"
-
-def quotedCharCoreFn (isQuotable : Char → Bool) : ParserFn := fun c s =>
-  let input := c.input
-  let i     := s.pos
-  if h : input.atEnd i then s.mkEOIError
-  else
-    let curr := input.get' i h
-    if isQuotable curr then
-      s.next' input i h
-    else if curr == 'x' then
-      andthenFn hexDigitFn hexDigitFn c (s.next' input i h)
-    else if curr == 'u' then
-      andthenFn hexDigitFn (andthenFn hexDigitFn (andthenFn hexDigitFn hexDigitFn)) c (s.next' input i h)
-    else
-      s.mkUnexpectedError "invalid escape sequence"
-
-def isQuotableCharDefault (c : Char) : Bool :=
-  c == '\\' || c == '\"' || c == '\'' || c == 'r' || c == 'n' || c == 't'
-
-def quotedCharFn : ParserFn :=
-  quotedCharCoreFn isQuotableCharDefault
-
-/-- Push `(Syntax.node tk <new-atom>)` onto syntax stack if parse was successful. -/
-def mkNodeToken (n : SyntaxNodeKind) (startPos : String.Pos) : ParserFn := fun c s => Id.run do
+/-- Function to be called directly after parsing a token. When not in an error state, parses following whitespace,
+    sets up `SourceInfo`, and pushes result of calling `f` with token substring and info onto stack. -/
+@[specialize]
+def pushToken (f : Substring → SourceInfo → Syntax) (startPos : String.Pos) (whitespaceFn : TokenParserFn) :
```

Comment:
I want to discuss this one. I see the motivations for the change, but the parser is starting to accumulate "patchwork."